File: lib/services/movepool.dart
import 'dart:math';
import '../models/move.dart';

/// Static table of the moves available for each of the 11 types in
/// the game, split into 3 power "tiers". Tier 1 is what starting
/// moves are drawn from (level 5), tier 2 unlocks around mid-career,
/// tier 3 for well-trained Wildkin.
///
/// NOTE: this is an MVP version with a handful of moves per type,
/// meant to be extended easily by adding entries to the list. In
/// production it's worth moving this table into a Postgres table
/// ('moves') so it can grow without app releases.
///
/// All move names below are original — not translations of any
/// existing game's move names — by design.
class MovePool {
  static const Map<String, List<Move>> _byType = {
    'fire': [
      Move(name: 'Spark Flame', type: 'fire', category: MoveCategory.special, power: 40, accuracy: 100, maxPp: 25, tier: 1),
      Move(name: 'Cinder Bite', type: 'fire', category: MoveCategory.physical, power: 45, accuracy: 95, maxPp: 20, tier: 1),
      Move(name: 'Blaze Jet', type: 'fire', category: MoveCategory.special, power: 90, accuracy: 100, maxPp: 15, tier: 2),
      Move(name: 'Smoldering Curse', type: 'fire', category: MoveCategory.status, power: 0, accuracy: 85, maxPp: 15, tier: 2),
      Move(name: 'Volcanic Burst', type: 'fire', category: MoveCategory.special, power: 120, accuracy: 100, maxPp: 5, tier: 3),
    ],
    'water': [
      Move(name: 'Water Jet', type: 'water', category: MoveCategory.special, power: 40, accuracy: 100, maxPp: 25, tier: 1),
      Move(name: 'Foam Burst', type: 'water', category: MoveCategory.special, power: 40, accuracy: 100, maxPp: 30, tier: 1),
      Move(name: 'Deluge Blast', type: 'water', category: MoveCategory.special, power: 90, accuracy: 90, maxPp: 10, tier: 2),
      Move(name: 'Tidal Wave', type: 'water', category: MoveCategory.special, power: 95, accuracy: 100, maxPp: 15, tier: 2),
      Move(name: 'Maelstrom', type: 'water', category: MoveCategory.special, power: 120, accuracy: 100, maxPp: 5, tier: 3),
    ],
    'grass': [
      Move(name: 'Bramble Snap', type: 'grass', category: MoveCategory.physical, power: 45, accuracy: 100, maxPp: 25, tier: 1),
      Move(name: 'Sap Drain', type: 'grass', category: MoveCategory.special, power: 20, accuracy: 100, maxPp: 25, tier: 1),
      Move(name: 'Thorned Slash', type: 'grass', category: MoveCategory.physical, power: 90, accuracy: 100, maxPp: 15, tier: 2),
      Move(name: 'Radiant Bloom', type: 'grass', category: MoveCategory.special, power: 120, accuracy: 100, maxPp: 10, tier: 3),
    ],
    'electric': [
      Move(name: 'Static Zap', type: 'electric', category: MoveCategory.special, power: 40, accuracy: 100, maxPp: 30, tier: 1),
      Move(name: 'Faint Spark', type: 'electric', category: MoveCategory.special, power: 45, accuracy: 100, maxPp: 25, tier: 1),
      Move(name: 'Voltaic Arc', type: 'electric', category: MoveCategory.special, power: 90, accuracy: 100, maxPp: 15, tier: 2),
      Move(name: 'Storm Bolt', type: 'electric', category: MoveCategory.special, power: 110, accuracy: 70, maxPp: 10, tier: 3),
    ],
    'ice': [
      Move(name: 'Chill Gust', type: 'ice', category: MoveCategory.special, power: 55, accuracy: 95, maxPp: 15, tier: 1),
      Move(name: 'Frost Orb', type: 'ice', category: MoveCategory.special, power: 40, accuracy: 90, maxPp: 25, tier: 1),
      Move(name: 'Glacial Ray', type: 'ice', category: MoveCategory.special, power: 90, accuracy: 100, maxPp: 10, tier: 2),
      Move(name: 'Frost Squall', type: 'ice', category: MoveCategory.special, power: 110, accuracy: 70, maxPp: 5, tier: 3),
    ],
    'poison': [
      Move(name: 'Toxic Soot', type: 'poison', category: MoveCategory.special, power: 40, accuracy: 100, maxPp: 25, tier: 1),
      Move(name: 'Venom Prick', type: 'poison', category: MoveCategory.physical, power: 15, accuracy: 100, maxPp: 35, tier: 1),
      Move(name: 'Corrosive Flame', type: 'poison', category: MoveCategory.special, power: 90, accuracy: 100, maxPp: 10, tier: 2),
      Move(name: 'Acid Surge', type: 'poison', category: MoveCategory.special, power: 100, accuracy: 90, maxPp: 10, tier: 3),
    ],
    'ground': [
      Move(name: 'Pitfall Strike', type: 'ground', category: MoveCategory.physical, power: 45, accuracy: 100, maxPp: 25, tier: 1),
      Move(name: 'Sand Toss', type: 'ground', category: MoveCategory.status, power: 0, accuracy: 100, maxPp: 15, tier: 1),
      Move(name: 'Tremor Quake', type: 'ground', category: MoveCategory.physical, power: 100, accuracy: 100, maxPp: 10, tier: 2),
      Move(name: 'Rising Dust', type: 'ground', category: MoveCategory.physical, power: 90, accuracy: 85, maxPp: 10, tier: 3),
    ],
    'flying': [
      Move(name: 'Wing Jab', type: 'flying', category: MoveCategory.physical, power: 35, accuracy: 100, maxPp: 35, tier: 1),
      Move(name: 'Wind Spiral', type: 'flying', category: MoveCategory.special, power: 40, accuracy: 100, maxPp: 25, tier: 1),
      Move(name: 'Sky Acrobat', type: 'flying', category: MoveCategory.physical, power: 85, accuracy: 100, maxPp: 15, tier: 2),
      Move(name: 'Cyclone Force', type: 'flying', category: MoveCategory.special, power: 110, accuracy: 70, maxPp: 10, tier: 3),
    ],
    'psychic': [
      Move(name: 'Mind Ripple', type: 'psychic', category: MoveCategory.special, power: 50, accuracy: 100, maxPp: 25, tier: 1),
      Move(name: 'Keen Insight', type: 'psychic', category: MoveCategory.special, power: 40, accuracy: 100, maxPp: 20, tier: 1),
      Move(name: 'Mental Shockwave', type: 'psychic', category: MoveCategory.special, power: 80, accuracy: 100, maxPp: 10, tier: 2),
      Move(name: 'Telekinetic Wave', type: 'psychic', category: MoveCategory.special, power: 90, accuracy: 100, maxPp: 10, tier: 3),
    ],
    'rock': [
      Move(name: 'Boulder Toss', type: 'rock', category: MoveCategory.physical, power: 50, accuracy: 90, maxPp: 15, tier: 1),
      Move(name: 'Riverstone Tackle', type: 'rock', category: MoveCategory.physical, power: 45, accuracy: 95, maxPp: 20, tier: 1),
      Move(name: 'Stone Cleaver', type: 'rock', category: MoveCategory.physical, power: 100, accuracy: 80, maxPp: 5, tier: 2),
      Move(name: 'Landslide', type: 'rock', category: MoveCategory.physical, power: 75, accuracy: 90, maxPp: 10, tier: 3),
    ],
    'dark': [
      Move(name: 'Swift Bite', type: 'dark', category: MoveCategory.physical, power: 40, accuracy: 100, maxPp: 25, tier: 1),
      Move(name: 'Shadowed Glare', type: 'dark', category: MoveCategory.status, power: 0, accuracy: 100, maxPp: 30, tier: 1),
      Move(name: 'Sneak Strike', type: 'dark', category: MoveCategory.physical, power: 40, accuracy: 100, maxPp: 30, tier: 2),
      Move(name: 'Grim Maw', type: 'dark', category: MoveCategory.physical, power: 80, accuracy: 100, maxPp: 15, tier: 3),
    ],
  };

  /// The 4 starting moves for a freshly captured Wildkin: always
  /// from tier 1, always matching its type(s).
  List<Move> starterMoves(List<String> types, {Random? random}) {
    final rng = random ?? Random();
    final pool = <Move>[];
    for (final type in types) {
      pool.addAll(_movesOfTier(type, 1));
    }
    // If the pool has fewer than 4 moves, fill in with the first
    // type's tier-1 moves again to reach 4 candidates.
    if (types.isNotEmpty) pool.addAll(_movesOfTier(types.first, 1));

    pool.shuffle(rng);
    final unique = <String, Move>{};
    for (final move in pool) {
      unique[move.name] = move;
      if (unique.length == 4) break;
    }
    return unique.values.toList();
  }

  /// A stronger move to offer as a replacement, unlocked at the
  /// current level. Returns null if there's nothing new to learn at
  /// this level (called every N levels; the cadence is decided by
  /// the caller).
  Move? nextMoveToLearn({
    required List<String> types,
    required int level,
    required List<String> alreadyKnownNames,
    Random? random,
  }) {
    final rng = random ?? Random();
    final targetTier = level >= 60 ? 3 : (level >= 25 ? 2 : 1);

    final candidates = <Move>[];
    for (final type in types) {
      candidates.addAll(
        _movesOfTier(type, targetTier)
            .where((m) => !alreadyKnownNames.contains(m.name)),
      );
    }
    if (candidates.isEmpty) return null;
    candidates.shuffle(rng);
    return candidates.first;
  }

  List<Move> _movesOfTier(String type, int tier) {
    final all = _byType[type.toLowerCase()] ?? const [];
    return all.where((m) => m.tier == tier).toList();
  }
}

File: supabase/functions/_shared/movepool.ts
// Port of lib/services/movepool.dart. Same table, same keys
// (name/type/category/power/accuracy/max_pp/tier) expected by
// Move.fromJson on the Flutter side. Restricted to the 11 types
// currently in the game (see type_chart.ts). Only starterMoves() is
// needed here for capture (tier 1); tiers 2/3 stay ready for a
// future evolve-wildkin function.
//
// All move names below are original — not translations of any
// existing game's move names — by design, matching movepool.dart.

export type MoveCategory = "physical" | "special" | "status";

export interface MoveJson {
  name: string;
  type: string;
  category: MoveCategory;
  power: number;
  accuracy: number;
  max_pp: number;
  tier: number;
}

const BY_TYPE: Record<string, MoveJson[]> = {
  fire: [
    { name: "Spark Flame", type: "fire", category: "special", power: 40, accuracy: 100, max_pp: 25, tier: 1 },
    { name: "Cinder Bite", type: "fire", category: "physical", power: 45, accuracy: 95, max_pp: 20, tier: 1 },
    { name: "Blaze Jet", type: "fire", category: "special", power: 90, accuracy: 100, max_pp: 15, tier: 2 },
    { name: "Smoldering Curse", type: "fire", category: "status", power: 0, accuracy: 85, max_pp: 15, tier: 2 },
    { name: "Volcanic Burst", type: "fire", category: "special", power: 120, accuracy: 100, max_pp: 5, tier: 3 },
  ],
  water: [
    { name: "Water Jet", type: "water", category: "special", power: 40, accuracy: 100, max_pp: 25, tier: 1 },
    { name: "Foam Burst", type: "water", category: "special", power: 40, accuracy: 100, max_pp: 30, tier: 1 },
    { name: "Deluge Blast", type: "water", category: "special", power: 90, accuracy: 90, max_pp: 10, tier: 2 },
    { name: "Tidal Wave", type: "water", category: "special", power: 95, accuracy: 100, max_pp: 15, tier: 2 },
    { name: "Maelstrom", type: "water", category: "special", power: 120, accuracy: 100, max_pp: 5, tier: 3 },
  ],
  grass: [
    { name: "Bramble Snap", type: "grass", category: "physical", power: 45, accuracy: 100, max_pp: 25, tier: 1 },
    { name: "Sap Drain", type: "grass", category: "special", power: 20, accuracy: 100, max_pp: 25, tier: 1 },
    { name: "Thorned Slash", type: "grass", category: "physical", power: 90, accuracy: 100, max_pp: 15, tier: 2 },
    { name: "Radiant Bloom", type: "grass", category: "special", power: 120, accuracy: 100, max_pp: 10, tier: 3 },
  ],
  electric: [
    { name: "Static Zap", type: "electric", category: "special", power: 40, accuracy: 100, max_pp: 30, tier: 1 },
    { name: "Faint Spark", type: "electric", category: "special", power: 45, accuracy: 100, max_pp: 25, tier: 1 },
    { name: "Voltaic Arc", type: "electric", category: "special", power: 90, accuracy: 100, max_pp: 15, tier: 2 },
    { name: "Storm Bolt", type: "electric", category: "special", power: 110, accuracy: 70, max_pp: 10, tier: 3 },
  ],
  ice: [
    { name: "Chill Gust", type: "ice", category: "special", power: 55, accuracy: 95, max_pp: 15, tier: 1 },
    { name: "Frost Orb", type: "ice", category: "special", power: 40, accuracy: 90, max_pp: 25, tier: 1 },
    { name: "Glacial Ray", type: "ice", category: "special", power: 90, accuracy: 100, max_pp: 10, tier: 2 },
    { name: "Frost Squall", type: "ice", category: "special", power: 110, accuracy: 70, max_pp: 5, tier: 3 },
  ],
  poison: [
    { name: "Toxic Soot", type: "poison", category: "special", power: 40, accuracy: 100, max_pp: 25, tier: 1 },
    { name: "Venom Prick", type: "poison", category: "physical", power: 15, accuracy: 100, max_pp: 35, tier: 1 },
    { name: "Corrosive Flame", type: "poison", category: "special", power: 90, accuracy: 100, max_pp: 10, tier: 2 },
    { name: "Acid Surge", type: "poison", category: "special", power: 100, accuracy: 90, max_pp: 10, tier: 3 },
  ],
  ground: [
    { name: "Pitfall Strike", type: "ground", category: "physical", power: 45, accuracy: 100, max_pp: 25, tier: 1 },
    { name: "Sand Toss", type: "ground", category: "status", power: 0, accuracy: 100, max_pp: 15, tier: 1 },
    { name: "Tremor Quake", type: "ground", category: "physical", power: 100, accuracy: 100, max_pp: 10, tier: 2 },
    { name: "Rising Dust", type: "ground", category: "physical", power: 90, accuracy: 85, max_pp: 10, tier: 3 },
  ],
  flying: [
    { name: "Wing Jab", type: "flying", category: "physical", power: 35, accuracy: 100, max_pp: 35, tier: 1 },
    { name: "Wind Spiral", type: "flying", category: "special", power: 40, accuracy: 100, max_pp: 25, tier: 1 },
    { name: "Sky Acrobat", type: "flying", category: "physical", power: 85, accuracy: 100, max_pp: 15, tier: 2 },
    { name: "Cyclone Force", type: "flying", category: "special", power: 110, accuracy: 70, max_pp: 10, tier: 3 },
  ],
  psychic: [
    { name: "Mind Ripple", type: "psychic", category: "special", power: 50, accuracy: 100, max_pp: 25, tier: 1 },
    { name: "Keen Insight", type: "psychic", category: "special", power: 40, accuracy: 100, max_pp: 20, tier: 1 },
    { name: "Mental Shockwave", type: "psychic", category: "special", power: 80, accuracy: 100, max_pp: 10, tier: 2 },
    { name: "Telekinetic Wave", type: "psychic", category: "special", power: 90, accuracy: 100, max_pp: 10, tier: 3 },
  ],
  rock: [
    { name: "Boulder Toss", type: "rock", category: "physical", power: 50, accuracy: 90, max_pp: 15, tier: 1 },
    { name: "Riverstone Tackle", type: "rock", category: "physical", power: 45, accuracy: 95, max_pp: 20, tier: 1 },
    { name: "Stone Cleaver", type: "rock", category: "physical", power: 100, accuracy: 80, max_pp: 5, tier: 2 },
    { name: "Landslide", type: "rock", category: "physical", power: 75, accuracy: 90, max_pp: 10, tier: 3 },
  ],
  dark: [
    { name: "Swift Bite", type: "dark", category: "physical", power: 40, accuracy: 100, max_pp: 25, tier: 1 },
    { name: "Shadowed Glare", type: "dark", category: "status", power: 0, accuracy: 100, max_pp: 30, tier: 1 },
    { name: "Sneak Strike", type: "dark", category: "physical", power: 40, accuracy: 100, max_pp: 30, tier: 2 },
    { name: "Grim Maw", type: "dark", category: "physical", power: 80, accuracy: 100, max_pp: 15, tier: 3 },
  ],
};

function movesOfTier(type: string, tier: number): MoveJson[] {
  const all = BY_TYPE[type.toLowerCase()] ?? [];
  return all.filter((m) => m.tier === tier);
}

function shuffle<T>(arr: T[]): T[] {
  const copy = [...arr];
  for (let i = copy.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1));
    [copy[i], copy[j]] = [copy[j], copy[i]];
  }
  return copy;
}

/// The 4 starting moves for a freshly captured Wildkin: always from
/// tier 1, matching its type(s), falling back to the first type's
/// tier-1 moves if the pool has fewer than 4.
export function starterMoves(types: string[]): MoveJson[] {
  const pool: MoveJson[] = [];
  for (const type of types) {
    pool.push(...movesOfTier(type, 1));
  }
  if (types.length > 0) pool.push(...movesOfTier(types[0], 1));

  const shuffled = shuffle(pool);
  const unique = new Map<string, MoveJson>();
  for (const move of shuffled) {
    unique.set(move.name, move);
    if (unique.size === 4) break;
  }
  return [...unique.values()];
}
